subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1373 @@
+	}
+
+	data->if_state = IF_STATE_INIT;
+
+	return 0;
+}
+
+static void usbd_cdc_ncm_shutdown(struct usbd_class_data *const c_data)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *const data = dev->data;
+	struct usbd_cdc_ncm_desc *desc = data->desc;
+
+	desc->if0_ecm.iMACAddress = 0;
+	sys_dlist_remove(&data->mac_desc_data->node);
+}
+
+static void *usbd_cdc_ncm_get_desc(struct usbd_class_data *const c_data,
+				   const enum usbd_speed speed)
+{
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct cdc_ncm_eth_data *const data = dev->data;
+
+	if (speed == USBD_SPEED_HS) {
+		return data->hs_desc;
+	}
+
+	return data->fs_desc;
+}
+
+#define TX_ALLOC_TIMEOUT 50 /* ms */
+
+static struct net_buf *tx_allocator(k_timeout_t timeout, void *user_data)
+{
+	return net_buf_alloc((struct net_buf_pool *)user_data, timeout);
+}
+
+static int cdc_ncm_send(const struct device *dev, struct net_pkt *const pkt)
+{
+	struct cdc_ncm_eth_data *const data = dev->data;
+	struct usbd_class_data *c_data = data->c_data;
+	size_t len = net_pkt_get_len(pkt);
+	struct net_buf *buf, *tmp;
+	union send_ntb *ntb;
+
+	if (len > NET_ETH_MAX_FRAME_SIZE) {
+		LOG_WRN("Trying to send too large packet, drop");
+		return -ENOMEM;
+	}
+
+	if (!atomic_test_bit(&data->state, CDC_NCM_CLASS_ENABLED) ||
+	    !atomic_test_bit(&data->state, CDC_NCM_IFACE_UP)) {
+		LOG_DBG("Configuration is not enabled or interface not ready (%d / %d)",
+			atomic_test_bit(&data->state, CDC_NCM_CLASS_ENABLED),
+			atomic_test_bit(&data->state, CDC_NCM_IFACE_UP));
+		return -EACCES;
+	}
+
+	buf = cdc_ncm_buf_alloc(cdc_ncm_get_bulk_in(c_data));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate buffer");
+		return -ENOMEM;
+	}
+
+	ntb = (union send_ntb *)buf->data;
+
+	ntb->nth.dwSignature = sys_cpu_to_le32(NTH16_SIGNATURE);
+	ntb->nth.wHeaderLength = sys_cpu_to_le16(sizeof(struct nth16));
+	ntb->nth.wSequence = sys_cpu_to_le16(++data->tx_seq);
+	ntb->nth.wNdpIndex = sys_cpu_to_le16(sizeof(struct nth16));
+	ntb->ndp.dwSignature = sys_cpu_to_le32(NDP16_SIGNATURE_NCM0);
+	ntb->ndp.wLength = sys_cpu_to_le16(sizeof(struct ndp16) +
+					   (CDC_NCM_SEND_MAX_DATAGRAMS_PER_NTB + 1) *
+					   sizeof(struct ndp16_datagram));
+	ntb->ndp.wNextNdpIndex = 0;
+	ntb->ndp_datagram[0].wDatagramIndex =
+		sys_cpu_to_le16(sys_le16_to_cpu(ntb->nth.wHeaderLength) +
+				sys_le16_to_cpu(ntb->ndp.wLength));
+	ntb->ndp_datagram[0].wDatagramLength = sys_cpu_to_le16(len);
+	ntb->ndp_datagram[1].wDatagramIndex  = 0;
+	ntb->ndp_datagram[1].wDatagramLength = 0;
+	ntb->nth.wBlockLength = sys_cpu_to_le16(
+		sys_le16_to_cpu(ntb->ndp_datagram[0].wDatagramIndex) + len);
+
+	net_buf_add(buf, sys_le16_to_cpu(ntb->ndp_datagram[0].wDatagramIndex));
+
+	tmp = pkt->frags;
+	while (tmp) {
+		size_t copied;
+
+		copied = net_buf_append_bytes(buf, tmp->len, tmp->data,
+					      K_MSEC(TX_ALLOC_TIMEOUT),
+					      tx_allocator, &cdc_ncm_ep_pool);
```

Comment:
It looks like it is the last one to resolve, https://github.com/zephyrproject-rtos/zephyr/pull/79562#discussion_r1802738585

Reply:
Yep, that was unnecessary what I was proposing and is now fixed.